binder/src/main/java/io/grpc/binder/util/ServiceBinding.java

```diff
@@ -0,0 +1,213 @@
+
+  // States can only ever transition in one direction.
+  private enum State {
+    NOT_BINDING,
+    BINDING,
+    BOUND,
+    UNBOUND,
+  }
+
+  private final ComponentName targetComponent;
+  private final String bindAction;
+  private final int bindFlags;
+  private final Observer observer;
+  private final Executor mainThreadExecutor;
+
+  @GuardedBy("this")
+  private State state;
+
+  // The following fields are intentionally not guarded, since (aside from the constructor),
+  // they're only used on the main thread. The constructor contains a synchronized block
+  // to ensure there's a write barrier when these fields are first written.
+
+  @Nullable private Context sourceContext; // Only null in the unbound state.
+
+  private State reportedState;
+
+  @AnyThread
+  ServiceBinding(
+      Executor mainThreadExecutor,
+      Context sourceContext,
+      ComponentName targetComponent,
+      String bindAction,
+      int bindFlags,
+      Observer observer) {
+    // We need to synchronize here ensure other threads see all
+    // non-final fields initialized after the constructor.
+    synchronized (this) {
+      this.targetComponent = targetComponent;
+      this.bindAction = bindAction;
+      this.bindFlags = bindFlags;
+      this.observer = observer;
+      this.sourceContext = sourceContext;
+      this.mainThreadExecutor = mainThreadExecutor;
+      state = State.NOT_BINDING;
+      reportedState = State.NOT_BINDING;
+    }
+  }
+
+  @MainThread
+  private void notifyBound(IBinder binder) {
+    if (reportedState == State.NOT_BINDING) {
+      reportedState = State.BOUND;
+      logger.log(Level.FINEST, "notify bound - notifying");
+      observer.onBound(binder);
+    }
+  }
+
+  @MainThread
+  private void notifyUnbound(Status reason) {
+    logger.log(Level.FINEST, "notify unbound ", reason);
+    clearReferences();
+    if (reportedState != State.UNBOUND) {
+      reportedState = State.UNBOUND;
+      logger.log(Level.FINEST, "notify unbound - notifying");
+      observer.onUnbound(reason);
+    }
+  }
+
+  @AnyThread
+  @Override
+  public synchronized void bind() {
+    if (state == State.NOT_BINDING) {
+      state = State.BINDING;
+      Intent bindIntent = new Intent(bindAction);
+      bindIntent.setComponent(targetComponent);
+      Status bindResult = bindInternal(sourceContext, bindIntent, this, bindFlags);
+      if (!bindResult.isOk()) {
+        state = State.UNBOUND;
+        mainThreadExecutor.execute(() -> notifyUnbound(bindResult));
+      }
+    }
+  }
+
+  private static Status bindInternal(
+      Context context, Intent bindIntent, ServiceConnection conn, int flags) {
+    try {
+      if (!context.bindService(bindIntent, conn, flags)) {
+        return Status.UNIMPLEMENTED.withDescription(
+            "bindService(" + bindIntent + ") returned false");
+      }
+      return Status.OK;
+    } catch (SecurityException e) {
+      return Status.PERMISSION_DENIED.withCause(e);
```

Comment:
Go ahead and also include descriptions here and below. The cause isn't always logged and may be lost, so it is nice to have a semi-unique string to track down where the status was generated.

Reply:
Done.